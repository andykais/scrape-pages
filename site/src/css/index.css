:root {
  --primary: #78909c;
  --primary-dark: #4b636e;
  --secondary: #bcaaa4;
  --secondary-light: white;
  --flow-error: #c40;
  --font-size: 16px;
}

html,
body {
  font-size: var(--font-size);
  padding: 0;
  margin: 0;
  font-family: sans-serif;
  height: 100%;
}
body {
  display: flex;
  flex-direction: column;
}
h1,
h2,
h3,
h4,
h5 {
  margin: 0;
}
a {
  text-decoration: none;
  position: relative;
  color: white;
}
i {
  font-size: 26px;
  line-height: 20px;
}
button i {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

#container {
  display: flex;
  flex-grow: 1;
}

@import './navbar.css';
@import './user-input.css';
@import './flow-diagram.css';
