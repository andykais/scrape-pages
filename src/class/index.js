import fillInDefaults from '../configuration/fill-in-defaults'

class Scraper {
  constructor(scrapeConfig) {}

  _urlSaver = scraper => children => input => {}
  _scalarSaver = scraper => children => input => {}
  _identitySaver = scraper => children => input => {}
  _htmlParser = () => {}
  _jsonParser = () => {}
}

const walkConfigCreateScrapers = config => {
  if (config.scrape_each) {
    walkConfigCreateScrapers(config.scrape_each)
  } else {
    return
  }
}

class ScrapePages {
  constructor(config) {
    this.config = fillInDefaults(config)
    this.requiredInput = this.config.input
    this.scrapers = walkConfigCreateScrapers(this.config.scrape)
  }

  scrape = () => {}
}

export default ScrapePages
