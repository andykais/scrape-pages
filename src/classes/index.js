import EventEmitter from 'events'
import Logger from '../logger'
import fillInDefaults from '../configuration/fill-in-defaults'
import chooseParser from './parsers'
import chooseSaver from './savers'
import { mkdirp } from '../util/fs-promise'
import * as Rx from 'rxjs'
import * as ops from 'rxjs/operators'
import { takeWhileHardStop } from '../rxjs-operators'
import { assertConfigType } from '../configuration/assert-type'
import Store from '../store'

export class Scraper {
  // constructor({ name, parse, download, scrapeEach }, io) {
  constructor(config, io) {
    const { name, parse, download, scrapeEach } = config
    const childless = !Boolean(scrapeEach.length)
    const { expect } = parse || {}

    this.name = name
    this.save = chooseSaver({ config, expect, ...io })
    this.parse = chooseParser({ config, expect, ...io })
    this.emitter = io.emitter
    this.logger = io.logger
    this.children = scrapeEach.map(scrape => new Scraper(scrape, io))
  }

  runSetup = async options => {
    await mkdirp(`${options.folder}/${this.name}`)
    await Promise.all(this.children.map(child => child.runSetup(options)))
  }

  // TODO recursively get operators instead of recusive run
  // then make flat observable
  //
  // TODO allow for increments like range(0, 100) where some may respond with nothing
  run = params => (parentValue, parentIndexes = []) =>
    this.save
      .run(params, parentIndexes)(parentValue)
      .pipe(
        ops.map(this.parse.run(params)),
        takeWhileHardStop(parsed => parsed.length),
        ops.mergeMap((parsed, incrementIndex) =>
          Rx.from(parsed).pipe(
            ops.mergeMap((value, parsedIndex) =>
              this.children.map(child =>
                child.run(params)(value, [
                  ...parentIndexes,
                  incrementIndex,
                  parsedIndex
                ])
              )
            )
          )
        ),
        ops.mergeAll()
      )
}

class ScrapePages {
  constructor(config) {
    assertConfigType(config)
    this.config = fillInDefaults(config)
    this.emitter = new EventEmitter() // dependency inject?
    this.logger = new Logger({ log_level: 3 })
    this.scrapingScheme = new Scraper(this.config.scrape, {
      emitter: this.emitter,
      logger: this.logger
    })
  }

  // TODO add parsable input for this first parse step
  run = async (input = {}, options = {}) => {
    this.logger.info('Begin downloading with inputs', input)
    try {
      if (!options.folder) throw new Error('need a download dir! (for now)')
      if (!this.isValidInput(input)) throw new Error('invalid input')
      await mkdirp(options.folder)
      await this.scrapingScheme.runSetup(options)

      this.scrapingScheme
        .run({ input, options })()
        .subscribe(
          () => {},
          error => {
            this.logger.error(error.toString())
            this.emitter.emit('error', error)
          },
          () => {
            // TODO add timer to show how long it took
            this.logger.info('Done!')
            this.emitter.emit('done')
          }
        )
      return this.emitter
    } catch (e) {
      this.logger.error(e)
      process.exit(1)
    }
  }

  isValidInput = input => {
    if (typeof this.config.input === 'string') {
      const inputKeyIsPresent = input[this.config.input]
      if (!inputKeyIsPresent) return false
    } else if (Array.isArray(this.config.input)) {
      const allInputsArePresent = Object.keys(this.config.input).every(
        key => input[key]
      )
      if (!allInputsArePresent) return false
    }
    return true
  }
}

export default ScrapePages
