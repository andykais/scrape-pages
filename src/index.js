import PageScraper from './classes'
import sourceMapSupport from 'source-map-support'
sourceMapSupport.install()

// TODO move this into a try catch around `run`
// I shouldnt be blocking unhandledRejections here
process.on('unhandledRejection', error => {
  console.error(error)
  process.exit(1)
})

export default PageScraper
