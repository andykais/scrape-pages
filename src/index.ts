export { scrape } from './scraper'

export * from './util/errors'

export * from './util'
