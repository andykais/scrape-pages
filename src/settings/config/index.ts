export { normalizeConfig } from './normalize'

export { flattenConfig } from './flatten'

export { assertConfigType } from './runtime/settings/config/assert'
