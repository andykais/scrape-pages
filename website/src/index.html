<!DOCTYPE html>
<html>
  <head>
    <meta name="viewport" content="initial-scale=1.0, width=device-width">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.2.0/css/all.css" integrity="sha384-hWVjflwFxL6sNzntih27bfxkr27PmbbK/iSvJ+a4+0owXq79v+lsFkW54bOGbiDQ" crossorigin="anonymous">
  </head>
  <body>
    <div id="navbar">
      <div id="title-container">
        <h2>scrape-pages</h2>
        <h5>config editor</h5>
      </div>
      <div class="icons">
        <span id="share">
          <i class="fas fa-share-alt"></i>
        </span>
        <a href="https://github.com/andykais/scrape-pages">
          <i class="fab fa-github"></i>
        </a>
      </div>
    </div>
    <div id="container">
      <div id="user-input">
        <textarea id="config-input" name="config-input" cols="30" rows="10"></textarea>
        <button id="normalize-button">
          <span>
            Normalize Config
          </span>
        </button>
      </div>
      <div id='hide-button-container'>
        <button id="hide-button">
          <i class="fas fa-caret-left"></i>
        </button>
      </div>
      <div id="config-dependent-content">
        <div id="tree-container"></div>
        <div id="flow-error-block"></div>
      </div>
    </div>
  </body>
</html>
