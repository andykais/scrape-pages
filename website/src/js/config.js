const config = {
  input: [],
  scrape: {
    name: "mainpage",
    download: {
      increment: 0,
      initialIndex: 0,
      headerTemplates: {},
      cookieTemplates: {},
      method: "GET",
      urlTemplate: "mainpage"
    },
    parse: { expect: "html", selector: "batchID" },
    scrapeEach: [
      {
        name: "grid",
        download: {
          increment: 1,
          initialIndex: 0,
          headerTemplates: {},
          cookieTemplates: {},
          method: "GET",
          urlTemplate: "batchPage{batchID}"
        },
        scrapeNext: {
          name: 'next-batchid',
          parse: {
            selector: "batchID",
            expect: "html"
          }
        },
        scrapeEach: [
          {
            name: "tag",
            parse: { expect: "html", selector: "span.tag" },
            scrapeEach: []
          },
          {
            name: "parse-image-page",
            parse: { expect: "html", selector: "li a.images" },
            scrapeEach: [
              {
                name: "image-page",
                download: {
                  increment: 0,
                  initialIndex: 0,
                  headerTemplates: {},
                  cookieTemplates: {},
                  method: "GET",
                  urlTemplate: "imagepage{imageID}"
                },
                parse: { expect: "html", selector: "img" },
                scrapeEach: [
                  {
                    name: "image",
                    download: {
                      increment: 0,
                      initialIndex: 0,
                      headerTemplates: {},
                      cookieTemplates: {},
                      method: "GET",
                      urlTemplate: "{src}"
                    },
                    scrapeEach: []
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
};
