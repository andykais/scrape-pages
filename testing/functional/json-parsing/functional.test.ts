import * as path from 'path'
import { expect } from 'chai'

import { RUN_OUTPUT_FOLDER, NockFolderMock } from '../../setup'
import { config, configParseJsonTwice, configParseJsonInsideScript } from './config'
import { scrape } from '../../../src'

const resourceFolder = `${__dirname}/fixtures`
const resourceUrl = `http://${path.basename(__dirname)}.com`

const options = {}
const params = {
  folder: path.resolve(RUN_OUTPUT_FOLDER, `${path.basename(__dirname)}`),
  cleanFolder: true
}

describe(__filename, () => {
  const siteMock = new NockFolderMock(resourceFolder, resourceUrl)

  beforeEach(siteMock.init)
  afterEach(siteMock.done)

  describe('with simple json response', () => {
    it('should get em', async () => {
      const { start, query } = scrape(config, options, params)
      const { on } = start()
      await new Promise(resolve => on('done', resolve))

      const result = query(['apiResponse'])
      expect(result.map(r => r['apiResponse'].map(c => c.parsedValue))).to.be.deep.equal([
        ['the', 'quick', 'brown', 'fox']
      ])
    })
  })

  describe('with json blob parsed nested', () => {
    it('should stringify, then parse, then stringify', async () => {
      const { start, query } = scrape(configParseJsonTwice, options, params)
      const { on } = start()
      await new Promise(resolve => on('done', resolve))

      const result = query(['parseContentFromPost'], { groupBy: 'parseContentFromPost' })
      expect(result).to.have.length(4)
      expect(result.map(r => r['parseContentFromPost'].map(c => c.parsedValue))).to.be.deep.equal([
        ['the'],
        ['quick'],
        ['brown'],
        ['fox']
      ])
    })
  })

  describe('with json that needs to be parsed out of a file', () => {
    it('should stringify, then parse, then stringify', async () => {
      const { start, query } = scrape(configParseJsonInsideScript, options, params)
      const { on } = start()
      await new Promise(resolve => on('done', resolve))

      const result = query(['jsonInJs'])
      expect(result[0]['jsonInJs']).to.have.length(9)
      expect(result.map(r => r['jsonInJs'].map(c => c.parsedValue))).to.be.deep.equal([
        ['the', 'quick', 'brown', 'fox', 'jumped', 'over', 'the', 'lazy', 'dog']
      ])
    })
  })
})
