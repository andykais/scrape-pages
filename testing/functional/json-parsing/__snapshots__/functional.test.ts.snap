// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing/functional/json-parsing/functional.test.ts with json blob parsed nested should stringify, then parse, then stringify 1`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "the",
      "scraper": "parseContentFromPost",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "quick",
      "scraper": "parseContentFromPost",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "brown",
      "scraper": "parseContentFromPost",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "fox",
      "scraper": "parseContentFromPost",
    },
  ],
]
`;

exports[`testing/functional/json-parsing/functional.test.ts with json that needs to be parsed out of a file should stringify, then parse, then stringify 1`] = `
Array [
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "the",
      "scraper": "parseContentFromPost",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "quick",
      "scraper": "parseContentFromPost",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "brown",
      "scraper": "parseContentFromPost",
    },
  ],
  Array [
    Object {
      "byteLength": null,
      "complete": 1,
      "downloadData": null,
      "filename": false,
      "id": true,
      "parsedValue": "fox",
      "scraper": "parseContentFromPost",
    },
  ],
]
`;

exports[`testing/functional/json-parsing/functional.test.ts with simple json response should get em 1`] = `
Array [
  Array [
    Object {
      "byteLength": "300.0",
      "complete": 1,
      "downloadData": "[\\"http://json-parsing.com/api-response.json\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": false,
      "id": true,
      "parsedValue": "the",
      "scraper": "apiResponse",
    },
    Object {
      "byteLength": "300.0",
      "complete": 1,
      "downloadData": "[\\"http://json-parsing.com/api-response.json\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": false,
      "id": true,
      "parsedValue": "quick",
      "scraper": "apiResponse",
    },
    Object {
      "byteLength": "300.0",
      "complete": 1,
      "downloadData": "[\\"http://json-parsing.com/api-response.json\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": false,
      "id": true,
      "parsedValue": "brown",
      "scraper": "apiResponse",
    },
    Object {
      "byteLength": "300.0",
      "complete": 1,
      "downloadData": "[\\"http://json-parsing.com/api-response.json\\",{\\"headers\\":{},\\"method\\":\\"GET\\"}]",
      "filename": false,
      "id": true,
      "parsedValue": "fox",
      "scraper": "apiResponse",
    },
  ],
]
`;
